### PR TITLE
[SPARK-12271][SQL] Improve error message when Dataset.as[ ] has incompatible schemas.

### PR DESCRIPTION
**Pull request by nongli**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -251,6 +251,27 @@ case class ExpressionEncoder[T](
...
+      // We couldn't resolve, this means something is wrong with the schemas.
+      analyzedPlan match {
+        case p @ Project(projectList, _) => {
+          p.projectList.foreach { e: NamedExpression => e.foreach { _ match {
+            case u @ UpCast(child, dataType, walkedTypePath) => {
+              throw new AnalysisException(s"Cannot resolve `${child.prettyString}`. " +
+                "The type path of the target object is:\n" +
+                walkedTypePath.mkString("", "\n", "\n") +
+                "Ensure that the input schema contains this field.\n - Input schema is " +
+                schema.map(_.simpleString).mkString(", "))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: should we also provide the type information of attributes in `schema`?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: the `simpleString` of attribute includes expr id, which I think is noisy to users, should we generate the string manually or use `StructType.fromAttributes` to turn `schema` into `StructType` and call `StrcutType.simpleString`?

**Comment by holdenk** [comment]
Could you maybe update the PR description to include the description from the JIRA?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 1 above)
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 2]
```diff
@@ -251,6 +251,25 @@ case class ExpressionEncoder[T](
 
     val plan = Project(Alias(unbound, "")() :: Nil, LocalRelation(schema))
     val analyzedPlan = SimpleAnalyzer.execute(plan)
+    if (!analyzedPlan.resolved) {
+      analyzedPlan match {
+        // Looked for UpCasts. If these still exist, it indicates they could not resolve and
+        // something about the schemas is mismatched.
+        case p @ Project(projectList, _) => {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Is there a reason to not just put this in `CheckAnalysis` and make it more general (i.e. not just `Project`)?  I realize its only used here now, but if we ever do use it elsewhere, its invalid to remain after analysis.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 3]
```diff
@@ -251,6 +251,25 @@ case class ExpressionEncoder[T](
...
+          p.projectList.foreach { e: NamedExpression => e.foreach { _ match {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: I think the extra `_ match {` is not needed.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 4]
```diff
@@ -481,10 +481,17 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
     val ds = Seq(2 -> 2.toByte, 3 -> 3.toByte).toDF("a", "b").as[ClassData]
     assert(ds.collect().toSeq == Seq(ClassData("2", 2), ClassData("3", 3)))
   }
-}
 
+  test("verify mismatching field names fail with a good error") {
+    val ds = Seq(ClassData("a", 1)).toDS()
+    intercept[AnalysisException] {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Check that the message has some key things in it? (variable names, types, etc.)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 2 above)
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: much better!

**Comment by marmbrus** [comment]
This looks great now!

LGTM pending tests / conflict resolution

**Comment by cloud-fan** [comment]
LGTM

**Comment by marmbrus** [comment]
Thanks, merging to master.

**Pull request closed without merging**